### ISSUE TITLE
Database migration error on fresh new deploy.

### ISSUE THREAD

**Issue by kukgini** [cloudfoundry-community/prometheus-boshrelease#115]
I'm deploying v18.5.0 on openstack with bosh cli v2 (AKA, usine menifests/prometheus.yml
Fresh deploy produce following error:
```
05:14:31 | Updating instance grafana: grafana/4a25aeae-9193-461b-bb13-5aea67d00d74 (0) (canary) (00:03:53)
            L Error: Action Failed get_task: Task aa9e4a77-d89e-4ae3-4138-22dd9d57b770 result: 1 of 1 post-start scripts failed. Failed Jobs: grafana.
```
            
/var/vcap/sys/log/grafana/post-start.stdout.log
```
t=2017-09-04T05:14:30+0000 lvl=info msg="Initializing DB" logger=sqlstore dbtype=postgres
t=2017-09-04T05:14:30+0000 lvl=info msg="Starting DB migration" logger=migrator
t=2017-09-04T05:14:30+0000 lvl=info msg="Executing migration" logger=migrator id="copy data account to org"
t=2017-09-04T05:14:30+0000 lvl=info msg="Skipping migration condition not fulfilled" logger=migrator id="copy data account to org"
t=2017-09-04T05:14:30+0000 lvl=info msg="Executing migration" logger=migrator id="copy data account_user to org_user"
t=2017-09-04T05:14:30+0000 lvl=info msg="Skipping migration condition not fulfilled" logger=migrator id="copy data account_user to org_user"
t=2017-09-04T05:14:30+0000 lvl=info msg="Executing migration" logger=migrator id="create annotation table v5"
t=2017-09-04T05:14:31+0000 lvl=eror msg="Executing migration failed" logger=migrator id="create annotation table v5" error="pq: duplicate key value violates unique constraint \"pg_type_typname_nsp_index\""
t=2017-09-04T05:14:31+0000 lvl=eror msg="Exec failed" logger=migrator error="pq: duplicate key value violates unique constraint \"pg_type_typname_nsp_index\"" sql="CREATE TABLE IF NOT EXISTS \"annotation\" (\n\"id\" SERIAL PRIMARY KEY  NOT NULL\n, \"org_id\" BIGINT NOT NULL\n, \"alert_id\" BIGINT NULL\n, \"user_id\" BIGINT NULL\n, \"dashboard_id\" BIGINT NULL\n, \"panel_id\" BIGINT NULL\n, \"category_id\" BIGINT NULL\n, \"type\" VARCHAR(25) NOT NULL\n, \"title\" TEXT NOT NULL\n, \"text\" TEXT NOT NULL\n, \"metric\" VARCHAR(255) NULL\n, \"prev_state\" VARCHAR(25) NOT NULL\n, \"new_state\" VARCHAR(25) NOT NULL\n, \"data\" TEXT NOT NULL\n, \"epoch\" BIGINT NOT NULL\n);"
t=2017-09-04T05:14:31+0000 lvl=eror msg="Fail to initialize orm engine" logger=sqlstore error="Sqlstore::Migration failed err: pq: duplicate key value violates unique constraint \"pg_type_typname_nsp_index\"\n"
```

**Comment by frodenas** [comment]
Interesting, I just deployed it on a new environment and didn't get that error. What `postgres` release are you using? 

Also, check the `/var/vcap/sys/log/grafana/post-start.std*.log`. The deployment seems to fail at the post-start script, not when stating grafana. Is there any error there? If yes, it's related to the database problem?

**Comment by kukgini** [comment]
postgres-9.6.4 in installed on 'database/0' vm.
It is what packaged in prometheus-release. 

/var/vcap/sys/log/grafana/post-start.stderr.log - there is no log (size is 0 kb)
/var/vcap/sys/log/grafana/post-start.stdout.log  - Above error is what I got.

Actually, database schema  is created. (i'm not sure it is ok). 
Next time when I re-deploy, post-start passed well.
Anyway, it is repetable. 

STEP:
1) delete-deployment (if already deployed)
2) deploy --> post-start failed
3) deploy again --> success

**Comment by kukgini** [comment]
Maybe this issue comes from my Openstack provider. (poor vm creation performance) 
I adjust to get deployment job in serial and increased canary & update watch time.
Then deployment successed. 

I guess 'bosh-agent' think 'post-start script' failed while it is still running and re-run another script again.
Because, parallel post-start script can produce"pq: duplicate key value violates unique constraint".

**Comment by frodenas** [comment]
Based on your explanation, I guess it's more about the `update_watch_time`.

`monit` starts the `grafana` process. While Grafana is performing the database migration, `monit` checks the process pid and concludes that it's running, and after the `update_watch_time` it declares it successful. Then `monit` starts the `post-start` script, but the Grafana process has not yet finished migrating the db. The `post-start` process then [resets the admin password](https://github.com/cloudfoundry-community/prometheus-boshrelease/blob/master/jobs/grafana/templates/bin/post-start#L19), which in fact, also tries to migrate the db. If both process migrates the db at the same time, then the `pq: duplicate key` problem arises.

But again, this is only a guess. I'll try to do some tests around this.

**Comment by frodenas** [comment]
This had been fixed at [v18.6.0](https://github.com/cloudfoundry-community/prometheus-boshrelease/releases/tag/v18.6.0). The [reset-admin-password](https://github.com/cloudfoundry-community/prometheus-boshrelease/blob/master/jobs/grafana/templates/bin/grafana-admin-password#L5) script now waits until `grafana` has migrated the db and is listening to the http port.

Thanks for reporting this bug!
